Custom repository for Scala artifacts not propagated to bloop

**Version(s)**
1.2.0

**Describe the bug**
In the Open Community Build we're building and publishing the Scala 3 compiler artefacts to the custom repository. These artifacts are used to check if projects can be build against given versions of the compiler. It's invoked like: 
```
scala-cli compile . --scala-version=3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT --repository=https://custom/maven/repo/3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT
``` 
When compiling with `--server=false` it works correctly, however, when running with bloop we get exception: 
```scala
Compiling project (Scala 3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT, JVM (11))
Error compiling project (Scala 3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT, JVM (11))
Error: Unexpected error when compiling repo_d49e65e4dd_6e665ca4ab: sbt.internal.inc.InvalidComponent: Couldn't retrieve module org.scala-lang:scala3-sbt-bridge:3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT:compile
        at sbt.internal.inc.BloopComponentCompiler.$anonfun$compileAndInstall$2(BloopComponentCompiler.scala:276)
        at sbt.internal.inc.BloopComponentCompiler.$anonfun$compileAndInstall$2$adapted(BloopComponentCompiler.scala:257)
        at sbt.io.IO$.withTemporaryDirectory(IO.scala:488)
        at sbt.io.IO$.withTemporaryDirectory(IO.scala:498)
        at sbt.internal.inc.BloopComponentCompiler.$anonfun$compileAndInstall$1(BloopComponentCompiler.scala:257)
        at sbt.internal.inc.BloopComponentCompiler.$anonfun$compileAndInstall$1$adapted(BloopComponentCompiler.scala:255)
        at sbt.io.IO$.withTemporaryDirectory(IO.scala:488)
        at sbt.io.IO$.withTemporaryDirectory(IO.scala:498)
        at sbt.internal.inc.BloopComponentCompiler.compileAndInstall(BloopComponentCompiler.scala:255)
        at sbt.internal.inc.BloopComponentCompiler.$anonfun$compiledBridgeJar$1(BloopComponentCompiler.scala:245)
        at sbt.internal.inc.IfMissing$Define.run(IfMissing.scala:19)
        at sbt.internal.inc.BloopComponentManager.createAndCache$1(BloopComponentManager.scala:46)
        at sbt.internal.inc.BloopComponentManager.$anonfun$files$3(BloopComponentManager.scala:60)
        at sbt.internal.inc.BloopComponentManager.getOrElse$1(BloopComponentManager.scala:38)
        at sbt.internal.inc.BloopComponentManager.$anonfun$files$2(BloopComponentManager.scala:60)
        at sbt.internal.inc.BloopComponentManager$$anon$2.call(BloopComponentManager.scala:89)
        at bloop.ComponentLock.apply(BloopComponentsLock.scala:10)
        at bloop.ComponentLock.apply$(BloopComponentsLock.scala:9)
        at bloop.BloopComponentsLock$.apply(BloopComponentsLock.scala:14)
        at sbt.internal.inc.BloopComponentManager.lock(BloopComponentManager.scala:89)
        at sbt.internal.inc.BloopComponentManager.$anonfun$lockSecondaryCache$1(BloopComponentManager.scala:84)
        at scala.Option.map(Option.scala:230)
        at sbt.internal.inc.BloopComponentManager.lockSecondaryCache(BloopComponentManager.scala:84)
        at sbt.internal.inc.BloopComponentManager.fromSecondary$1(BloopComponentManager.scala:58)
        at sbt.internal.inc.BloopComponentManager.$anonfun$files$6(BloopComponentManager.scala:64)
        at sbt.internal.inc.BloopComponentManager.getOrElse$1(BloopComponentManager.scala:38)
        at sbt.internal.inc.BloopComponentManager.$anonfun$files$5(BloopComponentManager.scala:64)
        at sbt.internal.inc.BloopComponentManager$$anon$2.call(BloopComponentManager.scala:89)
        at bloop.ComponentLock.apply(BloopComponentsLock.scala:10)
        at bloop.ComponentLock.apply$(BloopComponentsLock.scala:9)
        at bloop.BloopComponentsLock$.apply(BloopComponentsLock.scala:14)
        at sbt.internal.inc.BloopComponentManager.lock(BloopComponentManager.scala:89)
        at sbt.internal.inc.BloopComponentManager.lockLocalCache(BloopComponentManager.scala:82)
        at sbt.internal.inc.BloopComponentManager.files(BloopComponentManager.scala:64)
        at sbt.internal.inc.BloopComponentManager.file(BloopComponentManager.scala:72)
        at sbt.internal.inc.BloopComponentCompiler.compiledBridgeJar(BloopComponentCompiler.scala:245)
        at sbt.internal.inc.BloopComponentCompiler$BloopCompilerBridgeProvider.compiledBridge(BloopComponentCompiler.scala:123)
        at sbt.internal.inc.BloopComponentCompiler$BloopCompilerBridgeProvider.fetchCompiledBridge(BloopComponentCompiler.scala:129)
        at sbt.internal.inc.AnalyzingCompiler.getDualLoader(AnalyzingCompiler.scala:354)
        at sbt.internal.inc.AnalyzingCompiler.getCompilerLoader(AnalyzingCompiler.scala:343)
        at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:87)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.compileSources$1(BloopHighLevelCompiler.scala:133)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$7(BloopHighLevelCompiler.scala:155)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$1(BloopHighLevelCompiler.scala:71)
        at bloop.tracing.NoopTracer$.trace(BraveTracer.scala:53)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.timed$1(BloopHighLevelCompiler.scala:70)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$6(BloopHighLevelCompiler.scala:155)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at monix.eval.internal.TaskRunLoop$.startFull(TaskRunLoop.scala:81)
        at monix.eval.internal.TaskRestartCallback.syncOnSuccess(TaskRestartCallback.scala:101)
        at monix.eval.internal.TaskRestartCallback.onSuccess(TaskRestartCallback.scala:74)
        at monix.eval.internal.TaskExecuteOn$AsyncRegister$$anon$1.run(TaskExecuteOn.scala:71)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: coursierapi.error.SimpleResolutionError$1: Error downloading org.scala-lang:scala3-sbt-bridge:3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT
  not found: /Users/wmazur/.ivy2/local/org.scala-lang/scala3-sbt-bridge/3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT/ivys/ivy.xml
  not found: https://repo1.maven.org/maven2/org/scala-lang/scala3-sbt-bridge/3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT/scala3-sbt-bridge-3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT.pom
  not found: https://scala-ci.typesafe.com/artifactory/scala-integration/org/scala-lang/scala3-sbt-bridge/3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT/scala3-sbt-bridge-3.4.2-RC1-bin-8b1618573bdf0c9cd9704ed4772fcec9c213ca32-SNAPSHOT.pom
        at coursierapi.error.SimpleResolutionError.of(SimpleResolutionError.java:11)
        at coursierapi.shaded.coursier.internal.api.ApiHelper$.simpleResError(ApiHelper.scala:336)
        at coursierapi.shaded.coursier.internal.api.ApiHelper$.doFetch(ApiHelper.scala:380)
        at coursierapi.shaded.coursier.internal.api.ApiHelper.doFetch(ApiHelper.scala)
        at coursierapi.Fetch.fetchResult(Fetch.java:244)
        at coursierapi.Fetch.fetch(Fetch.java:239)
        at bloop.DependencyResolution$.resolveDependenciesWithErrors(DependencyResolution.scala:90)
        at bloop.DependencyResolution$.resolveWithErrors(DependencyResolution.scala:65)
        at sbt.internal.inc.BloopComponentCompiler.$anonfun$compileAndInstall$2(BloopComponentCompiler.scala:267)
        ... 58 more
```

Then investigating the logs you can see that custom repository `https://custom/maven/repo/...` was not included in the search of artifacts in the bloop

**To Reproduce**
Publish Scala 3 artifacts to custom remote repository, then try to use them in scala-cli

**Expected behaviour**
The custom repositories passed to the scala-cli should allow to find custom Scala 3 artefacts repositories.

--- a/modules/build/src/main/scala/scala/build/Build.scala
+++ b/modules/build/src/main/scala/scala/build/Build.scala
@@ -906,7 +906,8 @@ object Build {
           scalaVersion = params.scalaVersion,
           scalaBinaryVersion = params.scalaBinaryVersion,
           scalacOptions = scalacOptions.toSeq.map(_.value),
-          compilerClassPath = scalaArtifacts.compilerClassPath
+          compilerClassPath = scalaArtifacts.compilerClassPath,
+          bridgeJarsOpt = scalaArtifacts.bridgeJarsOpt
         )
         Some(compilerParams)
 

--- a/modules/build/src/main/scala/scala/build/Project.scala
+++ b/modules/build/src/main/scala/scala/build/Project.scala
@@ -48,7 +48,8 @@ final case class Project(
     val scalaConfigOpt = scalaCompiler.map { scalaCompiler0 =>
       bloopScalaConfig("org.scala-lang", "scala-compiler", scalaCompiler0.scalaVersion).copy(
         options = updateScalacOptions(scalaCompiler0.scalacOptions).map(_.value),
-        jars = scalaCompiler0.compilerClassPath.map(_.toNIO).toList
+        jars = scalaCompiler0.compilerClassPath.map(_.toNIO).toList,
+        bridgeJars = scalaCompiler0.bridgeJarsOpt.map(_.map(_.toNIO).toList)
       )
     }
     baseBloopProject(
@@ -229,6 +230,7 @@ object Project {
       options = Nil,
       jars = Nil,
       analysis = None,
-      setup = None
+      setup = None,
+      bridgeJars = None
     )
 }

--- a/modules/build/src/main/scala/scala/build/ScalaCompilerParams.scala
+++ b/modules/build/src/main/scala/scala/build/ScalaCompilerParams.scala
@@ -4,5 +4,6 @@ final case class ScalaCompilerParams(
   scalaVersion: String,
   scalaBinaryVersion: String,
   scalacOptions: Seq[String],
-  compilerClassPath: Seq[os.Path]
+  compilerClassPath: Seq[os.Path],
+  bridgeJarsOpt: Option[Seq[os.Path]]
 )

--- a/modules/cli/src/main/scala/scala/cli/commands/shared/SharedOptions.scala
+++ b/modules/cli/src/main/scala/scala/cli/commands/shared/SharedOptions.scala
@@ -438,7 +438,8 @@ final case class SharedOptions(
         python = sharedPython.python,
         pythonSetup = sharedPython.pythonSetup,
         scalaPyVersion = sharedPython.scalaPyVersion
-      )
+      ),
+      useBuildServer = compilationServer.server
     )
   }
 

--- a/modules/integration/src/test/scala/scala/cli/integration/SipScalaTests.scala
+++ b/modules/integration/src/test/scala/scala/cli/integration/SipScalaTests.scala
@@ -608,7 +608,7 @@ class SipScalaTests extends ScalaCliSuite with SbtTestHelper with MillTestHelper
           val localRepoPath = root / "local-repo"
           val sv            = "3.4.1-RC1"
           val artifactNames =
-            Seq("scala3-compiler_3", "scala3-sbt-bridge")
+            Seq("scala3-compiler_3") ++ (if (withBloop) Seq("scala3-sbt-bridge") else Nil)
           for { artifactName <- artifactNames } {
             val csRes = os.proc(
               TestUtil.cs,

--- a/modules/integration/src/test/scala/scala/cli/integration/SipScalaTests.scala
+++ b/modules/integration/src/test/scala/scala/cli/integration/SipScalaTests.scala
@@ -607,15 +607,19 @@ class SipScalaTests extends ScalaCliSuite with SbtTestHelper with MillTestHelper
         .fromRoot { root =>
           val localRepoPath = root / "local-repo"
           val sv            = "3.4.1-RC1"
-          val csRes = os.proc(
-            TestUtil.cs,
-            "fetch",
-            "--cache",
-            localRepoPath,
-            s"org.scala-lang:scala3-compiler_3:$sv"
-          )
-            .call(cwd = root)
-          expect(csRes.exitCode == 0)
+          val artifactNames =
+            Seq("scala3-compiler_3", "scala3-sbt-bridge")
+          for { artifactName <- artifactNames } {
+            val csRes = os.proc(
+              TestUtil.cs,
+              "fetch",
+              "--cache",
+              localRepoPath,
+              s"org.scala-lang:$artifactName:$sv"
+            )
+              .call(cwd = root)
+            expect(csRes.exitCode == 0)
+          }
           val buildServerOptions =
             if (withBloop) Nil else Seq("--server=false")
           val r = os.proc(

--- a/modules/options/src/main/scala/scala/build/Artifacts.scala
+++ b/modules/options/src/main/scala/scala/build/Artifacts.scala
@@ -104,6 +104,7 @@ object Artifacts {
     addJmhDependencies: Option[String],
     extraRepositories: Seq[Repository],
     keepResolution: Boolean,
+    includeBuildServerDeps: Boolean,
     cache: FileCache[Task],
     logger: Logger,
     maybeRecoverOnError: BuildException => Option[BuildException]
@@ -176,7 +177,9 @@ object Artifacts {
           ).left.flatMap(_.maybeRecoverWithDefault(Seq.empty, maybeRecoverOnError))
         }
 
-        val bridgeJarsOpt = Option.when(scalaArtifactsParams.params.scalaVersion.startsWith("3.")) {
+        val bridgeJarsOpt = Option.when(
+          scalaArtifactsParams.params.scalaVersion.startsWith("3.") && includeBuildServerDeps
+        ) {
           Seq(dep"org.scala-lang:scala3-sbt-bridge:${scalaArtifactsParams.params.scalaVersion}")
         }.map { bridgeDependencies =>
           value {

--- a/modules/options/src/main/scala/scala/build/Artifacts.scala
+++ b/modules/options/src/main/scala/scala/build/Artifacts.scala
@@ -176,6 +176,22 @@ object Artifacts {
           ).left.flatMap(_.maybeRecoverWithDefault(Seq.empty, maybeRecoverOnError))
         }
 
+        val bridgeJarsOpt = Option.when(scalaArtifactsParams.params.scalaVersion.startsWith("3.")) {
+          Seq(dep"org.scala-lang:scala3-sbt-bridge:${scalaArtifactsParams.params.scalaVersion}")
+        }.map { bridgeDependencies =>
+          value {
+            artifacts(
+              bridgeDependencies.map(Positioned.none),
+              allExtraRepositories,
+              Some(scalaArtifactsParams.params),
+              logger,
+              cache.withMessage(
+                s"Downloading Scala ${scalaArtifactsParams.params.scalaVersion} bridge"
+              )
+            ).left.flatMap(_.maybeRecoverWithDefault(Seq.empty, maybeRecoverOnError))
+          }.map(_._2)
+        }
+
         def fetchedArtifactToPath(fetched: Fetch.Result): Seq[os.Path] =
           fetched.fullDetailedArtifacts.collect { case (_, _, _, Some(f)) => os.Path(f, Os.pwd) }
 
@@ -273,7 +289,8 @@ object Artifacts {
           scalaNativeCli,
           internalDependencies,
           scalapyDependencies,
-          scalaArtifactsParams.params
+          scalaArtifactsParams.params,
+          bridgeJarsOpt
         )
         Some(scala)
 

--- a/modules/options/src/main/scala/scala/build/ScalaArtifacts.scala
+++ b/modules/options/src/main/scala/scala/build/ScalaArtifacts.scala
@@ -10,7 +10,8 @@ final case class ScalaArtifacts(
   scalaNativeCli: Seq[os.Path],
   internalDependencies: Seq[AnyDependency],
   extraDependencies: Seq[AnyDependency],
-  params: ScalaParameters
+  params: ScalaParameters,
+  bridgeJarsOpt: Option[Seq[os.Path]]
 ) {
 
   lazy val compilerClassPath: Seq[os.Path] =

--- a/modules/options/src/main/scala/scala/build/options/BuildOptions.scala
+++ b/modules/options/src/main/scala/scala/build/options/BuildOptions.scala
@@ -44,7 +44,8 @@ final case class BuildOptions(
   mainClass: Option[String] = None,
   testOptions: TestOptions = TestOptions(),
   notForBloopOptions: PostBuildOptions = PostBuildOptions(),
-  sourceGeneratorOptions: SourceGeneratorOptions = SourceGeneratorOptions()
+  sourceGeneratorOptions: SourceGeneratorOptions = SourceGeneratorOptions(),
+  useBuildServer: Option[Boolean] = None
 ) {
 
   import BuildOptions.JavaHomeInfo
@@ -457,6 +458,7 @@ final case class BuildOptions(
       addJmhDependencies = jmhOptions.addJmhDependencies,
       extraRepositories = value(finalRepositories),
       keepResolution = internal.keepResolution,
+      includeBuildServerDeps = useBuildServer.getOrElse(false),
       cache = finalCache,
       logger = logger,
       maybeRecoverOnError = maybeRecoverOnError

--- a/project/deps.sc
+++ b/project/deps.sc
@@ -115,7 +115,7 @@ object Deps {
   def ammonite = ivy"com.lihaoyi:::ammonite:3.0.0-M1-10-105f9e32"
   def asm      = ivy"org.ow2.asm:asm:9.7"
   // Force using of 2.13 - is there a better way?
-  def bloopConfig = ivy"ch.epfl.scala:bloop-config_2.13:1.5.5"
+  def bloopConfig = ivy"ch.epfl.scala:bloop-config_2.13:2.0.0"
     .exclude(("com.github.plokhotnyuk.jsoniter-scala", "jsoniter-scala-core_2.13"))
   def bloopRifle       = ivy"io.github.alexarchambault.bleep:bloop-rifle_2.13:${Versions.bloop}"
   def bsp4j            = ivy"ch.epfl.scala:bsp4j:2.1.1"

--- a/project/deps.sc
+++ b/project/deps.sc
@@ -107,7 +107,7 @@ object Deps {
     def signingCli                        = "0.2.3"
     def signingCliJvmVersion              = 17
     def javaClassName                     = "0.1.3"
-    def bloop                             = "1.5.16-sc-2"
+    def bloop                             = "1.5.17-sc-1"
   }
   // DO NOT hardcode a Scala version in this dependency string
   // This dependency is used to ensure that Ammonite is available for Scala versions

